File: packages/server/src/bootstrap/afd-context.ts
/**
 * @fileoverview afd-context bootstrap commands
 *
 * Context management commands for dynamic tool scoping.
 * Always available regardless of active context (universal tools).
 */

import type { CommandDefinition, CommandResult } from '@lushly-dev/afd-core';
import { failure, success } from '@lushly-dev/afd-core';
import { z } from 'zod';
import { resolveContextState } from '../context-scope.js';
import type { ContextConfig } from '../server-types.js';

// ═══════════════════════════════════════════════════════════════════════════════
// CONTEXT STATE
// ═══════════════════════════════════════════════════════════════════════════════

/** Default maximum depth of a context stack. */
export const DEFAULT_CONTEXT_MAX_DEPTH = 16;

export interface ContextState {
	stack: string[];
	/** Maximum stack depth (default: 16). `afd-context-enter` fails once it is reached. */
	maxDepth?: number;
	getActive(): string | null;
	enter(name: string): void;
	exit(): string | null;
}

/**
 * Create a context state manager: one stack of entered contexts.
 *
 * The MCP server keeps one for stdio and one per HTTP session. Re-entering the active context
 * is a no-op, and `enter` throws once `maxDepth` contexts are stacked.
 */
export function createContextState(options: { maxDepth?: number } = {}): ContextState {
	const maxDepth = options.maxDepth ?? DEFAULT_CONTEXT_MAX_DEPTH;
	if (!Number.isSafeInteger(maxDepth) || maxDepth <= 0) {
		throw new Error('Context maxDepth must be a positive integer');
	}
	const stack: string[] = [];

	return {
		stack,
		maxDepth,
		getActive() {
			return stack.length > 0 ? (stack[stack.length - 1] ?? null) : null;
		},
		enter(name: string) {
			if (stack[stack.length - 1] === name) return;
			if (stack.length >= maxDepth) {
				throw new Error(`Context stack is limited to ${maxDepth} levels`);
			}
			stack.push(name);
		},
		exit() {
			return stack.pop() ?? null;
		},
	};
}

/** Failure for context changes requested by a caller that has no session to hold them. */
function sessionRequired(): CommandResult<never> {
	return failure({
		code: 'SESSION_REQUIRED',
		message: 'Entering or exiting a context needs an MCP session, and this request has none',
		suggestion:
			'Send an MCP initialize request to /message and repeat its Mcp-Session-Id response header on later requests',
	});
}

// ═══════════════════════════════════════════════════════════════════════════════
// afd-context-list
// ═══════════════════════════════════════════════════════════════════════════════

const listInputSchema = z.object({});

interface ContextInfo {
	name: string;
	description?: string;
	triggers?: string[];
	priority?: number;
}

interface ContextListOutput {
	contexts: ContextInfo[];
	activeContext: string | null;
}

/**
 * Create the afd-context-list bootstrap command.
 */
export function createAfdContextListCommand(
	getContexts: () => ContextConfig[],
	contextState: ContextState
): CommandDefinition<z.infer<typeof listInputSchema>, ContextListOutput> {
	return {
		name: 'afd-context-list',
		description: 'List all configured contexts with descriptions, priorities, and triggers',
		category: 'bootstrap',
		tags: ['bootstrap', 'read', 'safe', 'context'],
		mutation: false,
		version: '1.0.0',
		parameters: [],

		async handler(_input, context) {
			const activeContext = resolveContextState(context, contextState)?.getActive() ?? null;
			const configs = getContexts();
			const contexts: ContextInfo[] = configs
				.sort((a, b) => (b.priority ?? 0) - (a.priority ?? 0))
				.map((c) => ({
					name: c.name,
					...(c.description && { description: c.description }),
					...(c.triggers?.length && { triggers: c.triggers }),
					...(c.priority != null && { priority: c.priority }),
				}));

			return success(
				{
					contexts,
					activeContext,
				},
				{
					reasoning: `Found ${contexts.length} configured contexts. Active: ${activeContext ?? 'none'}`,
					confidence: 1.0,
				}
			);
		},
	};
}

// ═══════════════════════════════════════════════════════════════════════════════
// afd-context-enter
// ═══════════════════════════════════════════════════════════════════════════════

const enterInputSchema = z.object({
	context: z.string().min(1).describe('Context name to enter'),
});

interface ContextEnterOutput {
	entered: string;
	previous: string | null;
}

/**
 * Create the afd-context-enter bootstrap command.
 */
export function createAfdContextEnterCommand(
	getContexts: () => ContextConfig[],
	contextState: ContextState
): CommandDefinition<z.infer<typeof enterInputSchema>, ContextEnterOutput> {
	return {
		name: 'afd-context-enter',
		description: 'Enter a context to scope available tools',
		category: 'bootstrap',
		tags: ['bootstrap', 'write', 'context'],
		mutation: true,
		version: '1.0.0',
		parameters: [
			{ name: 'context', type: 'string', required: true, description: 'Context name to enter' },
		],

		async handler(input: z.infer<typeof enterInputSchema>, context) {
			const state = resolveContextState(context, contextState);
			if (!state) return sessionRequired();
			const configs = getContexts();
			const contextName = input.context;

			const exists = configs.some((c) => c.name === contextName);
			if (!exists) {
				const available = configs.map((c) => c.name).join(', ');
				return failure({
					code: 'CONTEXT_NOT_FOUND',
					message: `Context '${contextName}' is not configured`,
					suggestion: `Available contexts: ${available}. Use afd-context-list to see all contexts.`,
				});
			}

			const previous = state.getActive();
			if (previous === contextName) {
				return success(
					{ entered: contextName, previous },
					{ reasoning: `Already in context '${contextName}'; nothing changed`, confidence: 1.0 }
				);
			}
			const maxDepth = state.maxDepth ?? DEFAULT_CONTEXT_MAX_DEPTH;
			if (state.stack.length >= maxDepth) {
				return failure({
					code: 'CONTEXT_DEPTH_EXCEEDED',
					message: `Context stack is limited to ${maxDepth} levels`,
					suggestion: 'Call afd-context-exit to leave a context before entering another',
				});
			}
			state.enter(contextName);

			return success(
				{ entered: contextName, previous },
				{
					reasoning: `Entered context '${contextName}'${previous ? ` (was '${previous}')` : ''}`,
					confidence: 1.0,
				}
			);
		},
	};
}

// ═══════════════════════════════════════════════════════════════════════════════
// afd-context-exit
// ═══════════════════════════════════════════════════════════════════════════════

const exitInputSchema = z.object({});

interface ContextExitOutput {
	exited: string | null;
	current: string | null;
}

/**
 * Create the afd-context-exit bootstrap command.
 */
export function createAfdContextExitCommand(
	contextState: ContextState
): CommandDefinition<z.infer<typeof exitInputSchema>, ContextExitOutput> {
	return {
		name: 'afd-context-exit',
		description: 'Exit the current context, popping back to the previous one',
		category: 'bootstrap',
		tags: ['bootstrap', 'write', 'context'],
		mutation: true,
		version: '1.0.0',
		parameters: [],

		async handler(_input, context) {
			const state = resolveContextState(context, contextState);
			if (!state) return sessionRequired();
			const exited = state.exit();
			const current = state.getActive();

			return success(
				{ exited, current },
				{
					reasoning: exited
						? `Exited context '${exited}'${current ? `, now in '${current}'` : ', no active context'}`
						: 'No active context to exit',
					confidence: 1.0,
				}
			);
		},
	};
}

File: python/src/afd/server/bootstrap/afd_context.py
"""afd-context bootstrap commands for dynamic tool scoping."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable, List, Optional

from pydantic import BaseModel, Field, ValidationError

from afd.core.commands import CommandContext, CommandDefinition, ExposeOptions
from afd.core.result import CommandResult, error, success
from afd.server.validation import _input_validation_failure


@dataclass(frozen=True)
class ContextConfig:
    """Configuration for a named tool context."""

    name: str
    description: Optional[str] = None
    triggers: List[str] = field(default_factory=list)
    priority: Optional[int] = None


class ContextState:
    """Mutable stack-based state for the currently active context."""

    def __init__(self) -> None:
        self.stack: List[str] = []

    def get_active(self) -> Optional[str]:
        return self.stack[-1] if self.stack else None

    def enter(self, name: str) -> None:
        self.stack.append(name)

    def exit(self) -> Optional[str]:
        return self.stack.pop() if self.stack else None


def create_context_state() -> ContextState:
    """Create a fresh mutable context state container."""
    return ContextState()


class AfdContextListInput(BaseModel):
    """Input for afd-context-list."""

    pass


class ContextInfo(BaseModel):
    """Metadata about a configured context."""

    name: str
    description: Optional[str] = None
    triggers: Optional[List[str]] = None
    priority: Optional[int] = None


class AfdContextListOutput(BaseModel):
    """Output for afd-context-list."""

    contexts: List[ContextInfo]
    active_context: Optional[str] = None


class AfdContextEnterInput(BaseModel):
    """Input for afd-context-enter."""

    context: str = Field(..., description="Context name to enter")


class AfdContextEnterOutput(BaseModel):
    """Output for afd-context-enter."""

    entered: str
    previous: Optional[str] = None


class AfdContextExitInput(BaseModel):
    """Input for afd-context-exit."""

    pass


class AfdContextExitOutput(BaseModel):
    """Output for afd-context-exit."""

    exited: Optional[str] = None
    current: Optional[str] = None


BOOTSTRAP_EXPOSE = ExposeOptions(mcp=True, cli=True)


def create_afd_context_list_command(
    get_contexts: Callable[[], List[ContextConfig]],
    context_state: ContextState,
) -> CommandDefinition:
    """Create the afd-context-list bootstrap command."""

    async def handler(
        input: Any,
        context: Optional[CommandContext] = None,
    ) -> CommandResult[AfdContextListOutput]:
        _ = context
        configs = sorted(
            get_contexts(),
            key=lambda item: item.priority or 0,
            reverse=True,
        )
        contexts = [
            ContextInfo(
                name=item.name,
                description=item.description,
                triggers=item.triggers or None,
                priority=item.priority,
            )
            for item in configs
        ]
        return success(
            AfdContextListOutput(
                contexts=contexts,
                active_context=context_state.get_active(),
            ),
            reasoning=f"Found {len(contexts)} configured contexts",
            confidence=1.0,
        )

    return CommandDefinition(
        name="afd-context-list",
        description="List all configured contexts with descriptions, priorities, and triggers",
        handler=handler,
        category="bootstrap",
        tags=["bootstrap", "read", "safe", "context"],
        mutation=False,
        version="1.0.0",
        input_schema=AfdContextListInput.model_json_schema(),
        returns=AfdContextListOutput.model_json_schema(),
        expose=BOOTSTRAP_EXPOSE,
    )


def create_afd_context_enter_command(
    get_contexts: Callable[[], List[ContextConfig]],
    context_state: ContextState,
) -> CommandDefinition:
    """Create the afd-context-enter bootstrap command."""

    async def handler(
        input: Any,
        context: Optional[CommandContext] = None,
    ) -> CommandResult[AfdContextEnterOutput]:
        _ = context
        try:
            parsed = (
                input
                if isinstance(input, AfdContextEnterInput)
                else AfdContextEnterInput.model_validate(input or {})
            )
        except ValidationError as exc:
            return _input_validation_failure(AfdContextEnterInput, input, exc)
        context_names = {item.name for item in get_contexts()}
        if parsed.context not in context_names:
            return error(
                "CONTEXT_NOT_FOUND",
                f"Context '{parsed.context}' is not configured",
                suggestion=(
                    "Use afd-context-list to see available contexts."
                ),
            )

        previous = context_state.get_active()
        context_state.enter(parsed.context)
        return success(
            AfdContextEnterOutput(entered=parsed.context, previous=previous),
            reasoning=f"Entered context '{parsed.context}'",
            confidence=1.0,
        )

    return CommandDefinition(
        name="afd-context-enter",
        description="Enter a context to scope available tools",
        handler=handler,
        category="bootstrap",
        tags=["bootstrap", "write", "context"],
        mutation=True,
        version="1.0.0",
        input_schema=AfdContextEnterInput.model_json_schema(),
        returns=AfdContextEnterOutput.model_json_schema(),
        expose=BOOTSTRAP_EXPOSE,
    )


def create_afd_context_exit_command(
    context_state: ContextState,
) -> CommandDefinition:
    """Create the afd-context-exit bootstrap command."""

    async def handler(
        input: Any,
        context: Optional[CommandContext] = None,
    ) -> CommandResult[AfdContextExitOutput]:
        _ = input, context
        exited = context_state.exit()
        current = context_state.get_active()
        return success(
            AfdContextExitOutput(exited=exited, current=current),
            reasoning=(
                f"Exited context '{exited}'"
                if exited
                else "No active context to exit"
            ),
            confidence=1.0,
        )

    return CommandDefinition(
        name="afd-context-exit",
        description="Exit the current context, popping back to the previous one",
        handler=handler,
        category="bootstrap",
        tags=["bootstrap", "write", "context"],
        mutation=True,
        version="1.0.0",
        input_schema=AfdContextExitInput.model_json_schema(),
        returns=AfdContextExitOutput.model_json_schema(),
        expose=BOOTSTRAP_EXPOSE,
    )
